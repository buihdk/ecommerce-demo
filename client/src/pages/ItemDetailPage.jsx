import React, { memo, useMemo, useState, useEffect } from 'react';
import { useParams } from 'react-router-dom';
import { get, find } from 'lodash';

import { fetchItemById } from 'services';
import { getCategories } from 'services/localStorage';
import ErrorBar from 'utils/ErrorBar';

import ItemDetail from 'components/ItemDetail';

export const useFetchItemById = (id, setRes) =>
  useEffect(() => {
    fetchItemById(id).then(response => setRes(response));
  }, [id, setRes]);

const ItemDetailPage = () => {
  const { id } = useParams();
  const [res, setRes] = useState({});
  const categories = useMemo(() => getCategories(), []);

  useFetchItemById(id, setRes);

  let categoryName;
  if (res && res.category_id) {
    categoryName = get(
      find(categories, category => Number(category.id) === res.category_id),
      'name',
      '',
    );
  }

  if (res && res.err) return <ErrorBar msg={res.err} />;
  return <ItemDetail item={res} categoryName={categoryName} />;
};

export default memo(ItemDetailPage);
