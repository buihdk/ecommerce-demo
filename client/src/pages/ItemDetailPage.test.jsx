import React from 'react';
import { act } from 'react-dom/test-utils';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { renderHook } from '@testing-library/react-hooks';

import ItemDetailPage, { useFetchItemById } from './ItemDetailPage';

const mockParams = { id: 1 };
jest.mock('react-router-dom', () => ({ useParams: () => mockParams }));

describe('ItemDetailPage', () => {
  const setRes = jest.fn();
  const fetchItemByIdSpy = jest.spyOn(require('../services'), 'fetchItemById');
  let wrapper;

  beforeEach(() => {
    fetchItemByIdSpy.mockReset();
  });

  test(`useFetchItems returns data`, async () => {
    fetchItemByIdSpy.mockImplementation(
      () =>
        new Promise(resolve => {
          resolve({ category_id: 1 });
        }),
    );
    await renderHook(() => useFetchItemById(1, setRes));
    await act(async () => {
      wrapper = shallow(<ItemDetailPage />);
    });

    expect(fetchItemByIdSpy).toBeCalled();
    expect(setRes).toBeCalledWith({ category_id: 1 });
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  test(`useFetchItems returns error`, async () => {
    fetchItemByIdSpy.mockImplementation(
      () =>
        new Promise(reject => {
          reject({ err: 'Error!' });
        }),
    );
    await renderHook(() => useFetchItemById(2, setRes));
    await act(async () => {
      wrapper = shallow(<ItemDetailPage />);
    });

    expect(fetchItemByIdSpy).toBeCalled();
    expect(setRes).toBeCalledWith({ err: 'Error!' });
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
