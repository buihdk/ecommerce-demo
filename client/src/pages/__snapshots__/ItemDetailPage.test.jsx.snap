// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemDetailPage useFetchItems returns data 1`] = `
<Memo(ItemDetail)
  item={Object {}}
/>
`;

exports[`ItemDetailPage useFetchItems returns error 1`] = `
<Memo(ItemDetail)
  item={Object {}}
/>
`;
